### ISSUE TITLE
Use nullable reference types throughout CSLA codebase

### ISSUE THREAD

**Issue by JasonBock** [MarimerLLC/csla#1233]
**Is your feature request related to a problem? Please describe.**
C#8 adds nullable reference types. Not only will CSLA need to support this, but it should also annotate its members with not just `?`, but with the new nullable attributes.

**Describe the solution you'd like**
NRT support throughout CSLA.

**Describe alternatives you've considered**
Nothing :)

**Additional context**
Use [this reference](https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/) for details/information.

**Comment by StefanOssendorf** [comment]
@rockfordlhotka
I'd love to tackle this issue. I don't like `null` :-)
Btw the nullable annotation context is not a breaking change. The annotation is only a compiler feature during development and not something the runtime checks.

**Comment by rockfordlhotka** [comment]
I'd love to have you do this!

It'll be CSLA 9 thing, as I'm sure there'll be breaking changes.

**Comment by StefanOssendorf** [comment]
@rockfordlhotka I'll link you here from now on for questions regarding the nullable context :-)
First question:
https://github.com/MarimerLLC/csla/blob/579c84dd14398549c1b48be3892d543a1a956260/Source/Csla/Core/FieldManager/FieldData.cs#L28-L42

Is the empty ctor only there for the `MobileFormatter`? It's at least not used within CSLA itself. I'd love to mark it `Obsolete(true)` because it would remove a lot of "maybe nullable" things in the class itself. For example the name can than be guaranteed.

**Comment by rockfordlhotka** [comment]
MobileFormatter does require an empty default ctor. Whether that is explicit or implicit, it is required for all types that implement `IMobileObject`.

**Comment by StefanOssendorf** [comment]
In that case I'll mark it as Obsolete. MobileFormatter can still use it but it's not usable any more at compile time :-)

**Comment by StefanOssendorf** [comment]
And should I _really_ add now everywhere a null argument check where it's currently just assumed to be not null? 😅🤔

**Comment by rockfordlhotka** [comment]
That seems wise doesn't it?

**Comment by JasonBock** [comment]
I concur with @rockfordlhotka. If a parameter is a NRT, someone could still pass in `null`, so an `ArgumentNullException.ThrowIfNull()` call is warranted.

**Comment by rockfordlhotka** [comment]
Hey - this is now part of the permanent record!! @JasonBock agrees with me!! 😁

**Comment by JasonBock** [comment]
@rockfordlhotka don't make me do it....
![image](https://github.com/MarimerLLC/csla/assets/904213/d09853c2-b156-47af-9dfc-02ca657d088a)

**Comment by StefanOssendorf** [comment]
@rockfordlhotka How should we handle `null` within the framework to e.g. reset/unset the `ClientContext`?
How do we handle things like this:
https://github.com/MarimerLLC/csla/blob/0f469350a0d72bf35d35deebb6660c6231c50b5d/Source/Csla/ApplicationContext.cs#L143-L147
I personally don't like to set something `null` when I just can assign a new instance to avoid that. At the cost of memory consumption obviously.
_But_ - and I think that's the biggest downside on allwoing `null` in such cases - we have to mark `ClientContext` as nullable which means every code says "Hey! This _could_ be `null` here. Please check that" eventhough _we know_ that won't *ever* be the case. But that's not expressable by the current annotations as far as I am aware.

**Comment by JasonBock** [comment]
If a property can be `null`, it needs to be annotated as such. And yes, that means that **every** usage of that property needs to do a null check. If the context should never be `null`, it should not be annotated with `?`, but, arguably, when you have a mutable property, you will need to do this regardless. The annotation is essentially a compilation helper, in that it can provide warnings to users to not do bad things with members that don't expect values to be `null`. But it doesn't stop it. You still have to check for `null`. And when the property is mutable, it means you **always** have to check it.

Consider this:

```c#
public sealed class Data
{
  public Data(string value)
  {
    ArgumentNullException.ThrowIfNull(value);
    this.Value = value;
  }

  public string Value { get; }
}
```

Since the property isn't settable after construction, there's no way it could be `null`. If it was, you would've received an exception during the creation of a `Data` instance. (Yes, you **could** do Reflection shenanigans to change that property value via the backing field, but I personally wouldn't be concerned with people digging themselves into holes this route). Therefore, I don't have to worry about checking `Value` all the time for a `null` value.

But if I do this:

```c#
public sealed class Data
{
  public Data(string value)
  {
    ArgumentNullException.ThrowIfNull(value);
    this.Value = value;
  }

  public string Value { get; set; }
}
```

Now I **have** to check for `null` every time I access `Value`. To keep the semantics of "`Value` will never be `null`", I would need to change the setter to do a `null` check, which would then ensure that `Value` wouldn't be `null`. 

Relating this back to `ClientContext`, either you annotate it with `?` because it's possible it could be `null`, or you don't, and then ensure that `ClientContext` can never be `null`. As you mentioned, this may mean you to provide an "empty" or "default" `ClientContext` instance. I don't know how that would work in CSLA. Or, since the current behavior is that it **could** be `null`, then put `?` on it, and make it explicit for callers.

**Comment by StefanOssendorf** [comment]
Of course if a 3rd party can _directly_ modifiy a property we have to make that nullable or change it to a backing field with a guard clause. In the case of `ClientContext` we have full control over it due to get/set methods where we can deny `null` as an acceptable value. The question is should we do that or not?
I personally would say: Yes, we should. It would not only make our code more robust and clean but our users one, too.

**Comment by JasonBock** [comment]
The current behavior states that `null` is a valid value, so it should be annotated with `?`. However, I'm guessing this will be part of a breaking change release, so maybe the right thing to do is to **not** annotate it, and force that either a non-`null` reference is provided, or an exception will occur. We know what will happen if we leave it nullable, because that's the current state of CSLA. Making it non-nullable may be the "right" change. I'm guessing @rockfordlhotka should weigh in on this.

**Comment by StefanOssendorf** [comment]
@JasonBock Yes. I'm asking because we are working on a breaking change version. Otherwise you are perfectly right. I would have annotatd it and keep the null behavior. Since I'm adding it _and_ we are in a breaking change version I'd like to make it a bit more nice imho :-)

**Comment by StefanOssendorf** [comment]
Thinking about it. The nullable feature would mean you can't use any auto properties because they don't adhere to the annotations.
There are cases even in the asp.net core framework like `HttpContext.Items` which are not nullable but the framework does not prevent you from setting it `null`.
So my stand will probably be: We use the feature so respect the feature. Otherwise garbage in -> garbage out :D

**Comment by rockfordlhotka** [comment]
The data portal has optimizations to minimize data on the wire through the use of null.

So if someone never uses a collection like ClientContext, then only a null token is serialized instead of the empty collection, which is much larger because it includes the type info.

As a result, anyone using a remote data portal is helped anytime a complex type has a null value instead of an instance of the type.

**Comment by rockfordlhotka** [comment]
When it comes to `ClientContext` (and `LocalContext`) the current implementation is to ensure that no one ever gets a null value, but _behind the scenes_ the value might be null so we get the serialization optimization.

This means that the _getter_ is what ensures that the value is never null. If the value is null, the setter creates it and returns the value.

That way, if someone doesn't use those context values (and this is a lot of users), then they pay no price for the feature existing in CSLA. But if someone _does_ use the context dictionaries, then they are created on demand.

As a result, in my view, the public API shouldn't change - people shouldn't worry that the hidden/private value behind the scenes could be null. We can change the private implementation if desired - annotating the _field_ with `?`, and the getter keeps doing what it does today.

I don't see any good reason to break everyone's code with a public API change. Either they use the context dictionaries and the dictionaries get lazy created, or they don't use them and the private field behind the scenes remains null.

**Comment by StefanOssendorf** [comment]
> The data portal has optimizations to minimize data on the wire through the use of null.
> 
> So if someone never uses a collection like ClientContext, then only a null token is serialized instead of the empty collection, which is much larger because it includes the type info.
> 
> As a result, anyone using a remote data portal is helped anytime a complex type has a null value instead of an instance of the type.

That's resonable. For me the correct place for such optimization would be the serializer and not the class to be serialized. But that's out of scope and maybe we'll tackle it at a later date.

**Comment by StefanOssendorf** [comment]
Please also take a look at `Csla.DataPortalClient.LocalProxyExtensions.UseLocalProxy(config, options)`. That implementation looks broken to me 🤔

**Comment by StefanOssendorf** [comment]
Please look at `Csla.Reflect.DynamicMethodHandle` ctor. The parameter `parameters` isn't used. Either that's a bug or it can be removed. For the latter I'll remove it :)

**Comment by rockfordlhotka** [comment]
> `DynamicMethodHandle`

It appears in use to me?

```csharp
        if (parameters == null)
        {
          inParams = [null];

        }
        else
        {
          inParams = parameters;
        }
```

**Comment by StefanOssendorf** [comment]
> > `DynamicMethodHandle`
> 
> It appears in use to me?
> 
> ```cs
>         if (parameters == null)
>         {
>           inParams = [null];
> 
>         }
>         else
>         {
>           inParams = parameters;
>         }
> ```

Yes. But `inParams` is not used. So when I remove `inParams` `parameters` is now unused, too.

**Comment by rockfordlhotka** [comment]
I think we need to create an issue to actually use the parameters. They are passed in when the ctor is called, and presumably are supposed to be used to ensure we use the right overload of the method.

**Comment by StefanOssendorf** [comment]
I'm now working on `Csla.Rules`.

**Comment by StefanOssendorf** [comment]
@rockfordlhotka I need some input please :)
We have these two rules:
`MinValue<T> where T: IComparable`
https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/CommonRules.cs#L371-L377
and
`MaxValue<T> where T: IComparable`
https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/CommonRules.cs#L459-L465

What would you expect to happen, when `value == null` in the line of comparison? That's a possibility there.
We can decide it counts as not violating the rule or we make the rule restriction tighter. For example T must be a struct.

**Comment by StefanOssendorf** [comment]
Another question: Any good reason why this lists can't just be initialized and hence empty if not neede and/or used? Makes the nullable stuff a lot easier and even more intuitive to use. Currently accessing `DirtyProperties` in a rule would return `null` instead of an empty list...
https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/RuleContext.cs#L84-L98
https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/RuleContext.cs#L66-L82

**Comment by rockfordlhotka** [comment]
> Another question: Any good reason why this lists can't just be initialized and hence empty if not neede and/or used? Makes the nullable stuff a lot easier and even more intuitive to use. Currently accessing `DirtyProperties` in a rule would return `null` instead of an empty list...
> 
> https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/RuleContext.cs#L84-L98
> 
> 
> https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/RuleContext.cs#L66-L82

CSLA often doesn't initialize lists of types that might be serialized, because it is _much_ smaller to serialize `null` than it is to serialize an empty list. This is because the serializer doesn't need to flow the list's type information for `null`, but it does need to flow the type information for an empty list so the empty list can be deserialized on the other end of the process.

**Comment by StefanOssendorf** [comment]
> > Another question: Any good reason why this lists can't just be initialized and hence empty if not neede and/or used? Makes the nullable stuff a lot easier and even more intuitive to use. Currently accessing `DirtyProperties` in a rule would return `null` instead of an empty list...
> > https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/RuleContext.cs#L84-L98
> > 
> > https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/RuleContext.cs#L66-L82
> 
> CSLA often doesn't initialize lists of types that might be serialized, because it is _much_ smaller to serialize `null` than it is to serialize an empty list. This is because the serializer doesn't need to flow the list's type information for `null`, but it does need to flow the type information for an empty list so the empty list can be deserialized on the other end of the process.

I understand that. But `List<>` isn't serializable and a rule context doesn't travel between client and server. So I think this should be okay here.

**Comment by StefanOssendorf** [comment]
> > @rockfordlhotka I need some input please :) We have these two rules: `MinValue<T> where T: IComparable`
> > https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/CommonRules.cs#L371-L377
> > 
> > and
> > `MaxValue<T> where T: IComparable`
> > https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/CommonRules.cs#L459-L465
> > 
> > What would you expect to happen, when `value == null` in the line of comparison? That's a possibility there. We can decide it counts as not violating the rule or we make the rule restriction tighter. For example T must be a struct.
> 
> I think in this context we should consider `null` to be `default(T)` for the purpose of comparison.

I don't understand your answer. My problem ist that `value` can be `null` which means I can't do `value.CompareTo()` because of a NullReferenceException.
How do we handle that?

**Comment by rockfordlhotka** [comment]
> I don't understand your answer. My problem ist that value can be null which means I can't do value.CompareTo() because of a NullReferenceException.

> How do we handle that?

**Comment by StefanOssendorf** [comment]
`ICslaIdentity`, `ICslaPrincipal`, `ICheckRoles.` also dead? 🤔

**Comment by StefanOssendorf** [comment]
Next is `Csla.Server`

**Comment by StefanOssendorf** [comment]
@rockfordlhotka 
* `Csla.DataPortalClient.DataPortalProxy.GetBaseCriteriaRequest()`
* `Csla.DataPortalClient.DataPortalProxy.GetBaseUpdateCriteriaRequest()`

Both set the `ClientCulture` and `ClientUICulture` property of the request object. But in different ways.
`System.Globalization.CultureInfo.CurrentCulture.Name` vs `Thread.CurrentThread.CurrentCulture.Name`.

Is that intended?

**Comment by StefanOssendorf** [comment]
Is the type `Csla.Server.MobileFactoryAttribute` obsolete? The documentation is talking about WCF and it's not used within the sources.

**Comment by StefanOssendorf** [comment]
https://github.com/MarimerLLC/csla/blob/e035ad787444a8a3f6b5df603d472030c124efc6/Source/Csla/Server/ObjectFactory.cs#L62-L75

Is it a bug that the `MethodCaller.CallMethodIfImplemented` part isn't trying to invoke the async version of check rules?

**Comment by rockfordlhotka** [comment]
Yes. That whole thing exists as a fallback for an edge case I suspect no one actually uses, but for completeness it should be made to work.

**Comment by StefanOssendorf** [comment]
> Yes. That whole thing exists as a fallback for an edge case I suspect no one actually uses, but for completeness it should be made to work.

That's a reply to which of my messages? :D The check rules or the attribute?

**Comment by rockfordlhotka** [comment]
> https://github.com/MarimerLLC/csla/blob/e035ad787444a8a3f6b5df603d472030c124efc6/Source/Csla/Server/ObjectFactory.cs#L62-L75
> 
> Is it a bug that the `MethodCaller.CallMethodIfImplemented` part isn't trying to invoke the async version of check rules?

Yes. That whole thing exists as a fallback for an edge case I suspect no one actually uses, but for completeness it should be made to work.

**Comment by StefanOssendorf** [comment]
> > https://github.com/MarimerLLC/csla/blob/e035ad787444a8a3f6b5df603d472030c124efc6/Source/Csla/Server/ObjectFactory.cs#L62-L75
> > 
> > Is it a bug that the `MethodCaller.CallMethodIfImplemented` part isn't trying to invoke the async version of check rules?
> 
> Yes. That whole thing exists as a fallback for an edge case I suspect no one actually uses, but for completeness it should be made to work.

Yeah. But shouldn't the fallback also tries to invoke `CheckRulesAsync` instead of just `CheckRules`?

**Comment by rockfordlhotka** [comment]
> > > https://github.com/MarimerLLC/csla/blob/e035ad787444a8a3f6b5df603d472030c124efc6/Source/Csla/Server/ObjectFactory.cs#L62-L75
> > > 
> > > Is it a bug that the `MethodCaller.CallMethodIfImplemented` part isn't trying to invoke the async version of check rules?
> > 
> > 
> > Yes. That whole thing exists as a fallback for an edge case I suspect no one actually uses, but for completeness it should be made to work.
> 
> Yeah. But shouldn't the fallback also tries to invoke `CheckRulesAsync` instead of just `CheckRules`?

Yes, that is what I'm saying. It should be made to work - which means trying to invoke both.

**Comment by rockfordlhotka** [comment]
> > > Another question: Any good reason why this lists can't just be initialized and hence empty if not neede and/or used? Makes the nullable stuff a lot easier and even more intuitive to use. Currently accessing `DirtyProperties` in a rule would return `null` instead of an empty list...
> > > https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/RuleContext.cs#L84-L98
> > > 
> > > https://github.com/MarimerLLC/csla/blob/aa14c3f86e3052d73cb06785b6fb6c381e6eb9f2/Source/Csla/Rules/RuleContext.cs#L66-L82
> > 
> > 
> > CSLA often doesn't initialize lists of types that might be serialized, because it is _much_ smaller to serialize `null` than it is to serialize an empty list. This is because the serializer doesn't need to flow the list's type information for `null`, but it does need to flow the type information for an empty list so the empty list can be deserialized on the other end of the process.
> 
> For `MobileFormatter`, yes, you need to serialize type information for a collection type so you know what to create on the other side. But that isn't a **strict** requirement for CSLA serialization. For example, my source generator serializer doesn't require this, because it knows at compile-time what the underlying type is for the field, and can `new` it up how it sees fit. For collections like `List<T>`, I don't special-case that one, but it's trivial to add a custom implementation for that in my generator through configuration.
> 
> Also, I do like empty collections over `null` in general overall. Then I don't need to wonder, "is this a `null` list or not?" everywhere in my code. If the performance implications of a `null` list over an empty one is significant during serialization, then....maybe keep them null, but if it's essentially negligible, then I'd **much rather** see an empty list.

I split this into its own issue (#4266) so we can discuss better.

**Comment by StefanOssendorf** [comment]
Is it intended that an aggregate exception populates an exception but a DataPortalException resumes the execution flow? 🤔
Line 131 vs 135
https://github.com/MarimerLLC/csla/blob/e035ad787444a8a3f6b5df603d472030c124efc6/Source/Csla/DataPortalT.cs#L119-L136

**Comment by rockfordlhotka** [comment]
> Is it intended that an aggregate exception populates an exception but a DataPortalException resumes the execution flow? 🤔 Line 131 vs 135
> 
> https://github.com/MarimerLLC/csla/blob/e035ad787444a8a3f6b5df603d472030c124efc6/Source/Csla/DataPortalT.cs#L119-L136

I am pretty sure this is intended.

If a `DataPortalException` is what occurs, that can safely flow up through the call stack. If an `AggregateException` occurs then that needs to be converted into a `DataPortalException` that can safely flow up through the call stack.

**Comment by StefanOssendorf** [comment]
But that's my question. A `DataPortalException` isn't bubbling up to the caller but catched and "handled". Where as the `AggregateException` packs it _into` a `DataPortalException` and throws that.

**Comment by rockfordlhotka** [comment]
> HandleCreateDataPortalException

I think the code is correct as-is. The `HandleDataPortalException` method exists to unpack and remove any top-level server-side `DataPortalException` before throwing the client-side `DataPortalException`. In the case of an `AggregateException` being caught directly on the client, if it isn't a server `DataPortalException`, then the exception can be directly wrapped in a client `DataPortalException` as it is thrown.

**Comment by StefanOssendorf** [comment]
Yep. You are totally right. I thought the ´Handle*` method is an extension point for users and resumes the execution flow afterwards. But it throws an exception 🤦🏻‍♂️

**Comment by StefanOssendorf** [comment]
Another look please :-)
https://github.com/MarimerLLC/csla/blob/6789bb52ce9a55130e961896a80641220f467e27/Source/Csla/DynamicListBase.cs#L266-L279
IMHO this is bad and a bug. I don't know if this base class is really in use.
`async void` for non event handler should be banned ^^'. If I read this right this will not bubble up any exceptions to the user during a save.

**Comment by rockfordlhotka** [comment]
> Another look please :-)
> 
> https://github.com/MarimerLLC/csla/blob/6789bb52ce9a55130e961896a80641220f467e27/Source/Csla/DynamicListBase.cs#L266-L279
> 
> 
> IMHO this is bad and a bug. I don't know if this base class is really in use.
> `async void` for non event handler should be banned ^^'. If I read this right this will not bubble up any exceptions to the user during a save.
> Edit: Here too:
> 
> https://github.com/MarimerLLC/csla/blob/6789bb52ce9a55130e961896a80641220f467e27/Source/Csla/DynamicListBase.cs#L346

I don't disagree in principal, but iirc this is required by Windows Forms data binding so I don't know that it can be fixed.

Hmm, maybe it can - is there also a `DynamicBindingListBase`? Yes there is. So we can probably break the API for `DynamicListBase` because it isn't limited to that particular data binding behavior.

**Comment by StefanOssendorf** [comment]
Hey everyone, I need some opinions on how to tackle NRTs :)

I've added NRT to the `Csla` project and am now reiterating it get rid of warnings produced during the change.

I have (currently) two major problems/concerns.

A) Annotating our generic mobile base classes.
As example `MobileDictionary`:

The declaration of that type must be `MobileDictionary<K, V> : Dictionary<K, V> where K: notnull` (I'm omitting everything that's not relevant for the NRT stuff)
The `SetChildren` method uses `SerializationInfo` as a parameter. And that parameter provides a method `T? GetValue<T>(string name)`.

And here's the problem:
`GetValue<T>` has to define the return type as `T?` because the current implementation (everything except relevant code pieces removed, even try-catch to keep it shorter)
```csharp
public T? GetValue<T>(string name)
{
  var value = _values[name].Value;
  return value != null ? Utilities.CoerceValue<T?>(value.GetType(), null, value) : (T?)value;
}
```
handles `null` explicitly so the compiler says: "You are possibly returning null so annotate the return type appropiately".

And now I'm stuck with `V` (from `MobileDictionary`) and `T?` to not match. Because `V` can be anything (nullable and non-nullable) but `T?` specifys it could be `null` even when `T` may not be nullable at all.

I currently see these options to "solve" that problem:
1. Just add `!` to every `GetValue` invocation and pray the data are correct. (Worst case `MobileDictionary<int, string>` and we add a null string which is not expected at all)
2. We have to adjust the Type annotations the whole call-chain down and replace a lot of "return null;" cases with "throw InvalidOperationException" to blow up and prevent wrong data
Maybe you have a better idea?

B)
The same holds for `P? GetProperty<P>()` which is maybe not expected. (Even more if we add the required property feature)
```csharp
protected P GetRequiredProperty<P>(PropertyInfo<P> propertyInfo)
{
  if (propertyInfo is null)
  {
    throw new ArgumentNullException(nameof(propertyInfo));
  }
  if (propertyInfo.RelationshipType.HasFlag(RelationshipTypes.LazyLoad))
  {
    throw new NotSupportedException("Lazy load is not allowed by an required property call. Use the normal GetProperty method.");
  }
  ThrowIfNotAllowedToRead();

  return ReadRequiredProperty<P>(propertyInfo);


  // Dummy to show the intention
  static void ThrowIfNotAllowedToRead()
  {
    throw new NotSupportedException();
  }
}

protected P ReadRequiredProperty<P>(IPropertyInfo propertyInfo)
{
  // Throw if property can not be read
  // return property value
}
```
I'm thinking about adding methods like that where I do not handle any `null`/`default` of a type explicitly and just throw an exception if I can't handle it. That would reduce a lot of `!` for us _and_ our users.

An example would be:
```csharp
public static readonly PropertyInfo<string> MyProperty = ...

public string My {
  get => GetProperty<string>(MyProperty)!; // <-- !-operator to suppress the (wrong) null annotation
  set => SetProperty<string>(MyProperty, value);
}
```

I hope this makes any sense to you guys at all. If not please just ask and I try to clarify as good as I can.

**Comment by kant2002** [comment]
The question, really - is possibility of null in this place avoidable, because of other reasons, that way slap `!`, or it's possible in some rare cases. 

There is third option - you also can silence warnings using 
```
public T GetValue<T>(string name)
{
  var value = _values[name].Value;
  Debug.Assert(value != null);
  return Utilities.CoerceValue<T?>(value.GetType(), null, value);
}
```

that way you preserve existing runtime behaviour where you don't have runtime penalty on checking for null and throw NRE in rare cases but in debug builds you can catch that. Also compiler would be happy

You maybe can utilize `[return: NotNullIfNotNull("value")]` on https://github.com/MarimerLLC/csla/blob/a1cebe05db1d828fbe33676af25ed7b2b71db2fd/Source/Csla/Utilities.cs#L153

More about it here https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis#postconditions-maybenull-and-notnull

**Comment by StefanOssendorf** [comment]
@rockfordlhotka 
Is there a reason why `SessionMessage` derives from `CommandBase<T>` instead of just being a `MobileObject`?

**Comment by rockfordlhotka** [comment]
I wanted managed properties, so `CommandBase` or `ReadOnlyBase` are the simplest ways to get that feature.

It could be a `MobileObject`. I'm not sure if that'd save any bytes over the wire - I don't think so?

**Comment by StefanOssendorf** [comment]
I don't know if it saves any bytes over the wire. But it feels like a misuse of a `CommandBase` type object. It's not created with an `IDataPortal` call which I would expect from any of these kind of types. So it should really only be either a `MobileObject` or it's own type with a custom serializer. We can now go fancy :D

So if you don't mind I will change it to a `MobileObject` instead. Which would make it easier usable in tests as well.

**Comment by rockfordlhotka** [comment]
That is fine with me.

**Comment by StefanOssendorf** [comment]
Just to be sure:
A criteria object is inteded to be instantiated with `new` and not through a dataporal, correct?

If yes, the `FieldManager` needs an `ApplicationContext` in it's constructor. But because the criteria isn't instantiated through a data portal it has none.
My current way around that is a new `bool` to indicate whether the current object type (e.g. Criteria) has the need of a fully functional `ApplicationContext` or not. Depending on that I've introduced a new `DummyApplicationContext` with a private constructor which is just there to satisfy the need of an `ApplicationContext`. But the instance itself is not capable of doing any work because it's not correctly instantiated.
That's hell of an ugly workaround but I currently do not see any other way without breaking all criteria users.
Tbh I'm inclined with the new way serialization works to make criterias obsolete and promote POCO types as criterias. Less code on our side and our users side and no dependencies on any framework stuff because it's a POCO.

What do you think about marking `CriteriaBase<T>` as obsolete (maybe in v9 already?). And we drop it in v11 or later so people have enough time to migrate to the new POCO serialization?

**Comment by hurcane** [comment]
Regarding, CriteriaBase, I recently migrated a large code base from CSLA 5 to 8. We ended up changing all of the CriteriaBase descendants to ReadOnlyBase and always use a DataPortal method.

Though somewhat painful to go through that change, I like the result because developers have fewer choices to make. Criteria classes are no longer special, and we follow the same rules regarding RunLocal or not when creating the object.

I support the idea of obsoleting the class for a future breaking change.

**Comment by rockfordlhotka** [comment]
I actually thought I'd marked `CriteriaBase` as obsolete already - apparently not?

Yes, we should mark it with a soft obsolete, because it shouldn't be needed any longer.
